### PR TITLE
fix(server) : fix unique property violation

### LINKED ISSUE #2814: 🐛 Bug Report: internal server error while trying to create a user with a username that already exisits
### What happened?

on create user, when I try to create a user with a username that already exists, I get an internal server error

### What you expected to happen

I expect to get something like  409 and a message that a user with the requested username already exists
https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#client_error_responses

image

### How to reproduce

generate an app and make a post request to create a user

### Amplication version

0.12.8

### Environment

_No response_

### Are you willing to submit PR?

Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by shubhwip**
Issue Number: closes #2814

## PR Details
 * fix(server) : fix unique property violation
 * return conflict http status
 * add new conflict error type
 * closes #2814
 

## PR Checklist
- [x] Tests for the changes have been added
- [ ] `npm test` doesn't throw any error

IMPORTANT: Please review the [CONTRIBUTING.md](https://github.com/amplication/amplication/blob/master/CODE_OF_CONDUCT.md) file for detailed contributing guidelines.

### CODE REVIEW

File: packages/amplication-data-service-generator/src/server/resource/controller/controller.base.template.ts [hunk 1]
```diff
@@ -11,6 +11,7 @@ import { Request } from "express";
 import { plainToClass } from "class-transformer";
 // @ts-ignore
 import { ApiNestedQuery } from "../../decorators/api-nested-query.decorator";
+import { Prisma } from "@amplication/prisma-db";
```
Comment (shubhwip) [inline packages/amplication-data-service-generator/src/server/resource/controller/controller.base.template.ts:14]: In generated app, this import is not working for me. what could be the reason ?

**Review by yuval-hazaz — COMMENTED** [review]
@shubhwip overall it looks good. but, I don't think we should use Prisma specific code in the controller. instead we better use an exceptionFilter and use it across all controllers and resolvers. 
have a look at this file from Amplication server for example
https://github.com/amplication/amplication/blob/master/packages/amplication-server/src/filters/GqlResolverExceptions.filter.ts

but, before you move forward with more changes, I would love to hear @arielweinberger thoughts about it

**Comment by arielweinberger** [comment]
I agree with @yuval-hazaz that we want to avoid Prisma-specific error handling in controllers. I'm okay with putting the logic in an exception filter.

Also, let's think about the error messages that propagate to the user. Currently it is a very verbose and not-so-helpful message - `There is a unique constraint violation, a new Resource cannot be created with this data`.

Can we make it more user friendly? For example, based on the entity type, to return a message along the lines of `This ${ENTITY} already exist`.

**Comment by shubhwip** [comment]
> I agree with @yuval-hazaz that we want to avoid Prisma-specific error handling in controllers. I'm okay with putting the logic in an exception filter.
> 
> Also, let's think about the error messages that propagate to the user. Currently it is a very verbose and not-so-helpful message - `There is a unique constraint violation, a new Resource cannot be created with this data`.
> 
> A conflict exception is 409, and generally we want `4XX` errors to be human readable and actionable.
> 
> Can we make it more user friendly? For example, based on the entity type, to return a message along the lines of `This ${ENTITY} already exist`.

`nestjs-prisma` provides a `PrismaClientExceptionFilter` to catch unhandled [PrismaClientKnownRequestError](https://www.prisma.io/docs/reference/api-reference/error-reference#prisma-client-query-engine) and returns different HttpStatus codes instead of 500 Internal server error. Should we use it directly or implement our own ?

**Comment by shubhwip** [comment]
> I agree with @yuval-hazaz that we want to avoid Prisma-specific error handling in controllers. I'm okay with putting the logic in an exception filter.
> 
> Also, let's think about the error messages that propagate to the user. Currently it is a very verbose and not-so-helpful message - `There is a unique constraint violation, a new Resource cannot be created with this data`.
> 
> A conflict exception is 409, and generally we want `4XX` errors to be human readable and actionable.
> 
> Can we make it more user friendly? For example, based on the entity type, to return a message along the lines of `This ${ENTITY} already exist`.

Now the response to the unique constraint violation looks like below
```
{
    "message": "Another record with the same key already exist (username)",
    "statusCode": 409
}
```

**Comment by souravjain540** [comment]
can we approve this? @arielweinberger

**Comment by shubhwip** [comment]
@yuval-hazaz @arielweinberger Please review.

**Comment by shubhwip** [comment]
On my local setup tests are failing for generated app and also facing other issue mentioned in discord
https://discord.com/channels/757179260417867879/781089015548870695/1015553670324162623
https://discord.com/channels/757179260417867879/781089015548870695/1015522020064231505

**Comment by yuval-hazaz** [comment]
@shubhwip we will review this soon, thanks for your patience

**Comment by yuval-hazaz** [comment]
@abrl91 @tupe12334 can one of you please test this PR so we can merge it ASAP?

**Comment by abrl91** [comment]
@shubhwip Can you please run on the `packages/amplication-data-service-generation` package the following command:
`npm run update-test-data-service-snapshot`

**Comment by abrl91** [comment]
run branch

**Comment by shubhwip** [comment]
> @shubhwip Can you please:
> 
> 1. merge master into your branch
> 2. accept the current snapshot files from the conflict files (the snapshot files from  branch mater)
> 3. run on the `packages/amplication-data-service-generation` package the following command:
>    `npm run update-test-data-service-snapshot`

Sure

**Comment by shubhwip** [comment]
> @shubhwip Can you please:
> 
> 1. merge master into your branch
> 2. accept the current snapshot files from the conflict files (the snapshot files from  branch mater)
> 3. run on the `packages/amplication-data-service-generation` package the following command:
>    `npm run update-test-data-service-snapshot`

Done @abrl91

**Comment by yuval-hazaz** [comment]
thanks @shubhwip 
The solution in general looks good. great work! 
we still need a core member to confirm that the functionality is tested and approved. We will do it soon.

Also, please see my comment about ts-ignore

**Comment by shubhwip** [comment]
> thanks @shubhwip The solution in general looks good. great work! we still need a core member to confirm that the functionality is tested and approved. We will do it soon.
> 
> Also, please see my comment about ts-ignore

Thank you @yuval-hazaz I am not able to see your comment.

**Review by yuval-hazaz — CHANGES_REQUESTED** [review]

File: packages/amplication-data-service-generator/src/server/static/src/filters/HttpExceptions.filter.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+      } else {
+        message =
+          `[${exception.code}]: ` +
+          this.exceptionShortMessage(exception.message);
+      }
+      if (!Object.keys(this.errorCodesStatusMapping).includes(exception.code)) {
+        return super.catch(exception, host);
+      }
+      const errorResponse = {
+        message: message,
```
Comment (abrl91) [inline packages/amplication-data-service-generator/src/server/static/src/filters/HttpExceptions.filter.ts:73]: The error message the user gets looks as follows:
```
{
  "errors": [
    {
      "message": "Http Exception",
      "extensions": {
        "code": "409",
        "response": {
          "statusCode": 409
        }
      }
    }
  ],
  "data": null
}
```
I think we should provide a more descriptive error message such as "another record with the requested username already exists"

Comment (shubhwip) [reply packages/amplication-data-service-generator/src/server/static/src/filters/HttpExceptions.filter.ts:73]: I tried with these changes and response looks like
```
{
    "message": "Another record with the same key already exist (username)",
    "statusCode": 409
}
```

File: packages/amplication-data-service-generator/src/server/static/src/filters/HttpExceptions.filter.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
+import {
+  ArgumentsHost,
+  Catch,
+  HttpException,
+  HttpServer,
+  HttpStatus,
+} from "@nestjs/common";
+import { BaseExceptionFilter } from "@nestjs/core";
+import {
+  // @ts-ignore
```
[the remark this answers is not in the document]
Comment (shubhwip) [reply packages/amplication-data-service-generator/src/server/static/src/filters/HttpExceptions.filter.ts:10]: `prisma import` is not available in the context of static files and that's why i'd to add @ts-ignore :).

**Comment by tupe12334** [comment]
@shubhwip Sorry for closing your PR accidentally. Can you reopen it?

**Comment by yuval-hazaz** [comment]
@abrl91 what is the current status of this PR?

**Comment by yuval-hazaz** [comment]
@abrl91 @tupe12334 @arielweinberger 
please have a look at this PR and let's either merge it or close it. it is open for too long

**Pull request merged**